lib/mutant/ast/structure.rb

```diff
@@ -0,0 +1,140 @@
+# frozen_string_literal: true
+
+module Mutant
+  module AST
+    # AST Structure metadata
+    module Structure
+      class Node
+        include Adamantium, Anima.new(:type, :fixed, :variable)
+
+        class Fixed
+          include Adamantium, Anima.new(:index, :name)
+
+          class Descendant < self
+          end
+
+          class Attribute < self
+          end
+
+          def attribute?
+            instance_of?(Attribute)
+          end
+
+          def descendant?
+            instance_of?(Descendant)
+          end
+
+          def value(node)
+            node.children.fetch(index)
+          end
+        end
+
+        def each_descendant_deep(node, &block)
+          each_descendant(node) do |descendant|
+            block.call(descendant)
+            Structure.for(descendant.type).each_descendant_deep(descendant, &block)
+          end
+        end
+
+        def each_node(node, &block)
+          block.call(node)
+          each_descendant_deep(node, &block)
+        end
+
+        def each_descendant(node, &block)
+          descendants.each_value do |child|
+            value = child.value(node)
+
+            block.call(value) if value
+          end
+
+          variable_descendants(node).each(&block)
+          self
+        end
+
+        def descendants
+          fixed.select(&:descendant?).to_h { |child| [child.name, child] }
+        end
+
+        def attributes
+          fixed.select(&:attribute?).to_h { |child| [child.name, child] }
+        end
+
+        def descendant(name)
+          maybe_descendant(name) or fail "Node #{type} does not have fixed descendant #{name}"
+        end
+
+        def attribute(name)
+          maybe_attribute(name) or fail "Node #{type} does not have fixed attribute #{name}"
+        end
+
+        def maybe_attribute(name)
+          attributes[name]
+        end
+
+        def maybe_descendant(name)
+          descendants[name]
+        end
+
+        def self.fixed(values)
+          values.each_with_index.map do |(klass, name), index|
+            klass.new(index: index, name: name)
+          end
+        end
+
+      private
+
+        def variable_descendants(node)
+          node.children[fixed.length..]
+        end
+      end
+
+      ALL = [
```

Comment:
Needs more extensions ;) But can replace lots of meta data in `Mutant::Muator::Node` and even unparser if we wanted to.